pkg/gadgets/socket-collector/tracer/tracer.go

```diff
@@ -0,0 +1,194 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License
+
+package tracer
+
+import (
+	"bufio"
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"net"
+	"runtime"
+
+	"github.com/cilium/ebpf/link"
+	gadgetv1alpha1 "github.com/kinvolk/inspektor-gadget/pkg/api/v1alpha1"
+	"github.com/vishvananda/netns"
+)
+
+//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -target bpfel -cc clang IterTCPv4 ./bpf/tcp4-collector.c -- -I../../.. -O2 -g -c -x c
+//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -target bpfel -cc clang IterUDPv4 ./bpf/udp4-collector.c -- -I../../.. -O2 -g -c -x c
+
+type Socket struct {
+	Protocol            string `json:"protocol"`
+	LocalAddress        string `json:"local_address"`
+	LocalPort           uint16 `json:"local_port"`
+	RemoteAddress       string `json:"remote_address"`
+	RemotePort          uint16 `json:"remote_port"`
+	Status              string `json:"status"`
+	KubernetesNamespace string `json:"kubernetes_namespace"`
+	KubernetesPod       string `json:"kubernetes_pod"`
+}
+
+func parseIPv4(ipU32 uint32) string {
+	ipBytes := make([]byte, 4)
+
+	binary.BigEndian.PutUint32(ipBytes, ipU32)
+	ip := net.IP(ipBytes)
+
+	return ip.String()
+}
+
+// Parse is done based on what is defined kernel: https://github.com/torvalds/linux/blob/v5.13/include/net/tcp_states.h#L12-L24
+func parseStatus(proto string, statusUint uint8) (string, error) {
+	statusMap := [...]string{"ESTABLISHED", "SYN_SENT", "SYN_RECV", "FIN_WAIT1", "FIN_WAIT2", "TIME_WAIT", "CLOSE", "CLOSE_WAIT", "LAST_ACK", "LISTEN", "CLOSING", "NEW_SYN_RECV"}
+
+	// Kernel enum starts from 1, adjust it to the statusMap
+	statusUint--
+
+	if len(statusMap) < int(statusUint) {
+		return "", fmt.Errorf("invalid %s status: %d", proto, statusUint+1)
+	}
+
+	status := statusMap[statusUint]
+
+	// Transform TCP status into something more suitable for UDP
+	if proto == "udp" {
+		if status == "ESTABLISHED" {
+			status = "ACTIVE"
+		} else if status == "CLOSE" {
+			status = "INACTIVE"
+		} else {
+			return "", fmt.Errorf("unexpected %s status %s", proto, status)
+		}
+	}
+
+	return status, nil
+}
+
+func getTCPIter() (*link.Iter, error) {
+	objs := IterTCPv4Objects{}
+	if err := LoadIterTCPv4Objects(&objs, nil); err != nil {
+		return nil, fmt.Errorf("failed to load TCP BPF objects: %v", err)
+	}
+	defer objs.Close()
+
+	it, err := link.AttachIter(link.IterOptions{
+		Program: objs.DumpTcp4,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to attach TCP BPF iterator: %v", err)
```

Comment:
General comment: errors should be wrapped with `%w` -> https://blog.golang.org/go1.13-errors#TOC_3.3. 

I think @marga-kinvolk meant to use `%v` when you're printing and **not** returning them.